ElectronClient/gui/NoteEditor/NoteBody/AceEditor/CodeMirror.tsx

```diff
@@ -0,0 +1,206 @@
+function CodeMirrorEditor(props: EditorProps, ref: any) {
+	const [editor, setEditor] = useState(null);
+	const [editorParent, setEditorParent] = useState(null);
+
+	// Codemirror plugins add new commands to codemirror (or change it's behavior)
+	// This command adds the smartListIndent function which will be bound to tab
+	useListIdent(CodeMirror);
+
+	const getScrollHeight = useCallback(() => {
+		if (editor) {
+			const info = editor.getScrollInfo();
+			const overdraw = editor.state.scrollPastEndPadding ? editor.state.scrollPastEndPadding : '0px';
+			return info.height - info.clientHeight - parseInt(overdraw);
+		}
+		return 0;
+	}, [editor]);
+
+	useEffect(() => {
+		if (props.cancelledKeys) {
+			for (let i = 0; i < props.cancelledKeys.mac.length; i++) {
+				const k = props.cancelledKeys.mac[i];
+				CodeMirror.keyMap.macDefault[k] = null;
+			}
+			for (let i = 0; i < props.cancelledKeys.default.length; i++) {
+				const k = props.cancelledKeys.default[i];
+				CodeMirror.keyMap.default[k] = null;
+			}
+		}
+	}, [props.cancelledKeys]);
+
+	useImperativeHandle(ref, () => {
+		return {
+			focus: () => {
+				editor.focus();
+			},
+			getSelection: () => {
+				return editor.getSelection();
+			},
+			getScrollHeight: () => {
+				return getScrollHeight();
+			},
+			getScrollPercent: () => {
+				if (editor) {
+					const info = editor.getScrollInfo();
+
+					return info.top / getScrollHeight();
+				}
+				return 0;
+			},
+			setScrollPercent: (p: number) => {
+				if (editor) {
+					editor.scrollTo(null, p * getScrollHeight());
+				}
+			},
+			somethingSelected: () => {
+				return editor.somethingSelected();
+			},
+			getSelections: () => {
+				return editor.getSelections();
+			},
+			listSelections: () => {
+				return editor.listSelections();
+			},
+			replaceSelections: (replacement: string[]) => {
+				editor.replaceSelections(replacement, 'around');
+			},
+			getCursor: () => {
+				return editor.getCursor('anchor');
+			},
+			insertAtCursor: (insert: string) => {
+				const pos = editor.getCursor('anchor');
+				editor.replaceRange(insert, pos);
+			},
+		};
+	});
+
+	const editor_change = useCallback((cm: any, change: any) => {
+		if (props.onChange && change.origin !== 'setValue') {
+			props.onChange(cm.getValue());
+		}
+	}, [props.onChange]);
+
+	// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
+	const editor_scroll = useCallback((_cm: any) => {
+		props.onScroll();
+	}, [props.onScroll]);
+
+	// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
+	const editor_mousedown = useCallback((_cm: any, event: any) => {
+		if (event && event.button === 2) {
+			props.onEditorContextMenu();
+		}
+	}, [props.onEditorContextMenu]);
+
+	// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
+	const editor_paste = useCallback((_cm: any, _event: any) => {
+		props.onEditorPaste();
+	}, [props.onEditorPaste]);
+
+	const divRef = useCallback(node => {
```

Comment:
For references, you should probably use `useRef`. Then to initialise the editor, you can use `useEffect(() => {}, [])`

Reply:
What's the benefit of this change? This function get's called with a reference to the div everytime it's loaded and unloaded, I otherwise don't need the reference. I'm happy to make the change, I'm just new to hooks and looking to learn it better.

Reply:
It's mostly to keep things more standard. I think the way you've done it, you've more or less reimplemented useRef.

Reply:
I see what you mean, I changed it in the way you suggested and it ended up being much cleaner. :+1: